### DISCUSSION TITLE
need help with mqtt subscription in a script

### DISCUSSION THREAD

**Discussion by SilentBob83** [arendst/Tasmota#13732] [Q&A] [answered]
Hello, 
i would like to do a very simple script.

Subscribe to a topic. Receive the published value and control a Power.
I have compiled with the option:

#define SUPPORT_MQTT_EVENT

in my script i have:

>S
if mqttc>0
then
+>Subscribe Switch_up,cmnd/%topic%,Switch_up
endif

console output:
MQT: stat/tasmota_171430/RESULT = {"Subscribe":"Switch_up,cmnd/tasmota_%06X/#,Switch_up"}

1. I wonder why the topic does not appear in my broker ?
2. I created manually the topic and set a value, but i see the following output in the console

MQT: stat/tasmota_171430/RESULT = {"Command":"Unknown"}

3. how can i access the value , once it is shown as received properly in the console

could please someone help me ?

**Comment by barbudor** [comment]
> `Subscribe Switch_up,cmnd/%topic%,Switch_up`

This is probably a bad idea as you are subscribing to `cmnd/%topic%/#`, your event will receive every command sent through MQTT to your device

What are you exactly trying to achieve ?
On which topic would you like to publish to create your event ?

> I wonder why the topic does not appear in my broker ?

Why would you anything appears on your broker until you publish to the topic ?
How are you check thing that by the way ? on which broker ?

> I created manually the topic and set a value, but i see the following output in the console

How ?
Please provide the exact command that you have typed in the console

> how can i access the value , once it is shown as received properly in the console

![image](https://user-images.githubusercontent.com/2996042/142690203-73100d33-634b-4159-aa3f-14cad4d9dec1.png)

**Comment by SilentBob83** [comment]
I would lile to subscribe to  Switch_up,cmnd/%topic%/Switch_up
I thought the key would be the final Topic Name.

I use the MQTT broker from IObroker.
In the past I programmed a similar sw as Tasmota on my own with the pubsubclient libary.
If you subscribe a topic there it will be displayed automatically in the broker

With IObroker I create a object in a tree view ...where every / is like a branch
![Unbenannt](https://user-images.githubusercontent.com/8195709/142691231-0a02f630-0cf6-43b1-9dbc-846ea0342d55.PNG)
in this case it will send a topic cmnd/tasmota_xxxx/Switch_up once i change the value of the object.

Now i tried  the following - this is how i interpret the description from you 

>D
Switch_up=0

>S
if mqttc>0
then
+>Subscribe Switch_up,cmnd/%topic%/Switch_up
endif

>E
=>print %Switch_up%

still the same result command unknown

**Comment by SilentBob83** [comment]
The target is to add a tasmote based switch to my shutter control... maybe there is a much better way to do so (eventually without MQTT)

**Comment by barbudor** [comment, marked as answer]
First let's get with Tasmota wording:
- A switch (or a button) is a input to Tasmota, like when you connect a physical switch to a GPIO to send orders or states to Tasmota. You don't end a command to a switch, you get states (or state changes) from it.
- A relay is an output to Tasmota, like when you drive a relay from a GPIO. You can control a relay output with tha POWER command (from either the console or through MQTT `cmnd/%topic%/POWER` topic.

By default when you have a switch (or a button) and a relay on the same Tasmota, the 1st control the 2nd so when you flip the switch or press the button, the relay output will change just thet same as if a POWER command was received.
If you don't want this to happen and manage yourself switch/button events and relay power control, you need to "decouple" the switch/button from the relay. Read the "Buttons and Switches" section of the doc as there are different ways of doing this dependinng on your exact goal.


Now, from what you are explaining, I feel that you don't need a "switch input" but a "relay output"
If you have a GPIO configured as a Relay, you can publish ON or OFF payload to topic `cmnd/%topic%/POWER` to control the relay. NO need for rules or script for that.

What you do in IObroker to be able to publish on a given topic is beyond Tasmota. I am not using IOBroker but Mosquitto which do not need any preliminary declaration of a topic. I am using NodeRed, command line tools or others to publish on topics.

**Comment by SilentBob83** [comment]
OK, i might found now a way to fix my problem. 
My situation is not so easy.

I cannot connect my switch directley to tasmota since i have only 2 wires with 230V ( one for up and one for down).
I will install two 230V trigger modules ( with optocoupler ).

I already did so but not close to the shutter controller. I installed it in the distribution box ( approx.10m away) . 
-> Here i got some induction since there are 230V wires running in paralell and this was causing uncontrolled recognition of impulses.

I hope that the installation of the optocoupler closer to the shutter controller will fix my problem.

I also wrote a script to read a MCP23017 and control the shutter. This is working on my desk. But i am wondering if there is a simple option to connect the MCP Intput to the shutters ?

**Reply by Jason2866** [reply]
Please do not hijack a answered thread with a not related question.
You probably get no answer and it is not searchable for all others.
Thx.

**Reply by barbudor** [reply]
@SilentBob83 all you are writing is not making much sense. Without diagrams it's hard to follow you.

**Comment by SilentBob83** [comment]
I am not allowed to write anything else to my whole problem in this thread as jason mentioned above

**Comment by Jason2866** [comment]
Open a new discussion and provide the infos as barbudor wrote.
